files/default/neutron-ha-tool.py

```diff
@@ -576,6 +600,200 @@ def l3_agent_evacuate(qclient, agent_host, agent_picker, router_filter,
+
+    target_list = list_alive_dhcp_agents_except(agent_list,
+                                                exclude_agent=agent_to_evacuate)
+
+    if len(target_list) < 1:
+        LOG.error("There are no DHCP agents alive to migrate networks onto")
+        return 0
+
+    (migrations, errors) = \
+        migrate_dhcp_networks_from_agent(qclient, agent_to_evacuate,
+                                      target_list, agent_picker,
+                                      noop, wait_for_network, fail_fast)
+    LOG.info("%d networks %s evacuated from DHCP agent %s", migrations,
+             "would have been" if noop else "were", agent_host)
+    if errors > 0:
+        LOG.error("%d errors encountered during evacuation", errors)
+
+    return errors
+
+
+def migrate_dhcp_networks_from_agent(qclient, agent, targets, agent_picker,
+                                     noop, wait_for_network, fail_fast):
+    LOG.info("Querying agent_id=%s for networks to migrate away", agent['id'])
+
+    networks_on_agent = qclient.list_networks_on_dhcp_agent(agent['id'])['networks']
+
+    agent_picker.set_agents(targets)
+
+    migrations = 0
+    errors = 0
+    for network in networks_on_agent:
+        target = agent_picker.pick()
+
+        if noop:
+            LOG.info("Would try to migrate network=%s from agent=%s "
+                 "to agent=%s", network['id'], agent['id'], target['id'])
+            migrations += 1
+            continue
+
+        try:
+            migrate_dhcp_network(qclient, network, agent, target, wait_for_network)
+            migrations += 1
+        except:
+            LOG.exception("Failed to migrate network_id=%s from agent=%s to "
+                          "agent=%s", network['id'], agent['id'], target['id'])
+            errors += 1
+            if fail_fast:
+                break
+
+    return (migrations, errors)
+
+
+def migrate_dhcp_network(qclient, network, agent, target,
+                         wait_for_network=True):
+    """
+    Returns nothing, and raises exceptions on errors.
+
+    :param qclient: A neutronclient
+    :param network: The network to migrate
+    :param agent: The DHCP agent to migrate from
+    :param target: The DHCP agent to migrate to
+    :param wait_for_network: Optional wait-for-dhcp-network flag
+    """
+
+    network_id = network['id']
+    agent_id = agent['id']
+    target_id = target['id']
+
+    LOG.info("Migrating network=%s from agent=%s to agent=%s",
+             network_id, agent_id, target_id)
+
+    # Prevent SIGTERM only when modifying data through the API, and allow it
+    # when ensuring the API operations completed successfully
+    with term_disabled():
+
+        try:
+            # Remove the network from the original agent
+            qclient.remove_network_from_dhcp_agent(agent_id, network_id)
+            LOG.debug("Removed network from agent=%s" % agent_id)
+        # network might have been removed in the mean time
+        except NotFound as e:
```

Comment:
I believe it's actually `ObjectNotFound`, see https://pastebin.nue.suse.com/25407/src (neutron-server log)

Reply:
fixed but not tested

Reply:
Sorry this was wrong